File: 03/solution.js
import {check, readInput} from "../Utils.js";

function part1(input) {
    const visited = new Set()
    visited.add("0,0")
    let currentX = 0
    let currentY = 0
    input[0].split('').forEach(direction => {
        switch (direction) {
            case '>': currentX++; break
            case '<': currentX--; break
            case 'v': currentY--; break
            case '^': currentY++; break
        }
        visited.add("" + currentX + "," + currentY)
    })
    return visited.size
}

function part2(input) {
    const visited = new Set()
    visited.add("0,0")
    const currentX = [0, 0]
    const currentY = [0, 0]
    let turn = 0
    input[0].split('').forEach(direction => {
        switch (direction) {
            case '>': currentX[turn]++; break
            case '<': currentX[turn]--; break
            case 'v': currentY[turn]--; break
            case '^': currentY[turn]++; break
        }
        visited.add("" + currentX[turn] + "," + currentY[turn])
        turn = (turn + 1) % 2
    })
    return visited.size
}

const testInput1 = readInput("test_input_1.txt")
const testInput2 = readInput("test_input_2.txt")
const testInput3 = readInput("test_input_3.txt")
const input = readInput("input.txt")

check(part1(testInput1), 2)
check(part1(testInput2), 4)
check(part1(testInput3), 2)
console.log("Part 1: " + part1(input))

check(part2(testInput1), 2)
check(part2(testInput2), 3)
check(part2(testInput3), 11)
console.log("Part 2: " + part2(input))

File: 03/solution.go
package main

import (
	utils "advent-of-code-2016"
	"fmt"
	"strconv"
	"strings"
)

func part1(input string) string {
	lines := utils.SplitIntoLines(input)
	count := 0
	for _, line := range lines {
		numbers := strings.Fields(line)
		a, _ := strconv.Atoi(numbers[0])
		b, _ := strconv.Atoi(numbers[1])
		c, _ := strconv.Atoi(numbers[2])
		if a+b > c && a+c > b && b+c > a {
			count++
		}
	}
	return strconv.Itoa(count)
}

func part2(input string) string {
	lines := utils.SplitIntoLines(input)
	count := 0
	for i := 0; i < len(lines); i += 3 {
		firstRow := strings.Fields(lines[i])
		secondRow := strings.Fields(lines[i+1])
		thirdRow := strings.Fields(lines[i+2])
		for j := 0; j < 3; j++ {
			a, _ := strconv.Atoi(firstRow[j])
			b, _ := strconv.Atoi(secondRow[j])
			c, _ := strconv.Atoi(thirdRow[j])
			if a+b > c && a+c > b && b+c > a {
				count++
			}
		}

	}
	return strconv.Itoa(count)
}

func main() {
	input := utils.ReadInput("03", "input.txt")
	fmt.Printf("Part 1: %s\n", part1(input))
	fmt.Printf("Part 2: %s\n", part2(input))
}
